x/poe/simulation/operations.go

```diff
@@ -0,0 +1,339 @@
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simtypes.Account, chainID string,
+	) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+
+		validators, _, err := k.ValsetContract(ctx).ListValidators(ctx, nil)
+		if len(validators) == 0 || err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgDelegate, "numbers of validator equal zero"), nil, nil
+		}
+
+		simAccount, _ := simtypes.RandomAcc(r, accs)
+
+		denom := k.GetBondDenom(ctx)
+		balance := bk.GetBalance(ctx, simAccount.Address, denom)
+		if balance.IsNil() || !balance.IsPositive() {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgDelegate, "balance is negative"), nil, nil
+		}
+
+		spendable := bk.SpendableCoins(ctx, simAccount.Address).AmountOf(denom)
+		if !spendable.IsPositive() {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgDelegate, "spendable coins amount is negative"), nil, nil
+		}
+
+		amount, err := simtypes.RandPositiveInt(r, spendable)
+		if err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgDelegate, "unable to generate positive amount"), nil, err
+		}
+
+		selfDelegation := sdk.NewCoin(denom, amount)
+
+		var fees sdk.Coins
+		coins, hasNeg := sdk.NewCoins(sdk.NewCoin(denom, spendable)).SafeSub(sdk.Coins{selfDelegation})
+		if !hasNeg {
+			fees, err = simtypes.RandomFees(r, ctx, coins)
+			if err != nil {
+				return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgDelegate, "unable to generate fees"), nil, err
+			}
+		}
+
+		msg := types.NewMsgDelegate(simAccount.Address, selfDelegation, sdk.NewCoin(denom, sdk.ZeroInt()))
+
+		txCtx := simulation.OperationInput{
+			App:           app,
+			TxGen:         simappparams.MakeTestEncodingConfig().TxConfig,
+			Cdc:           nil,
+			Msg:           msg,
+			MsgType:       msg.Type(),
+			Context:       ctx,
+			SimAccount:    simAccount,
+			AccountKeeper: ak,
+			ModuleName:    types.ModuleName,
+		}
+
+		return simulation.GenAndDeliverTx(txCtx, fees)
+	}
+}
+
+// SimulateMsgUndelegate generates a MsgUndelegate with random values
+func SimulateMsgUndelegate(bk types.XBankKeeper, ak types.AccountKeeper, k keeper.Keeper) simtypes.Operation {
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simtypes.Account, chainID string,
+	) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+		validators, _, err := k.ValsetContract(ctx).ListValidators(ctx, nil)
+		if len(validators) == 0 || err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "numbers of validator equal zero"), nil, nil
+		}
+
+		val := validators[rand.Intn(len(validators))]
+		valAddr, err := sdk.AccAddressFromBech32(val.OperatorAddress) // plain string to not run into bech32 prefix issues with valoper
+		if err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "validator address is not valid"), nil, err
+		}
```

Comment:
L270-279 look very familiar, now. Would it make sense to extract this to a helper function?